### PR TITLE
Add `F.diagonal`

### PR DESCRIPTION
**Pull request by toslunar**
This PR adds numpy-compatible (taking-)diagonal operation to `chainer.functions`.

### CODE REVIEW

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
The failed Jenkins testcase is `TestUnpoolingND_param_3.test_double_backward_cpu`.

File: tests/chainer_tests/functions_tests/array_tests/test_diagonal.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+
+    def check_double_backward(self, x_data, y_grad, x_grad_grad):
+        def f(x):
+            x = functions.diagonal(x, *self.args)
+            return x * x
+
+        gradient_check.check_double_backward(
+            f, x_data, y_grad, x_grad_grad,
+            **self.check_double_backward_options)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_diagonal.py]: Please use `dtype=numpy.float64` as raw argument.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by okuta** [comment]
LGTM

**Pull request merged**